src/solve/fulfill.rs

```diff
@@ -217,67 +241,119 @@ impl<'s> Fulfill<'s> {
+        // Attempt to refute each negative obligations by in a fresh Fulfill
+        for goal in self.to_refute.drain(..) {
+            let canonicalized = self.infer.canonicalize(&goal);
+            let goal = canonicalized.quantified.value;
+
+            // Negation cannot be used to resolve existential variables, and do
+            // not have a useful (for us) logial meaning when they contain
+            // existential variables; treat negative goals with free variables
+            // as ambiguous
+            if !canonicalized.free_vars.is_empty() {
+                return Ok(Outcome::Incomplete);
+            }
+
+
+            // Try to *solve* the goal in a fresh Fulfill
+            let mut fulfill = Fulfill::new(self.solver);
+            fulfill.push_goal(&goal.environment, goal.goal);
+
+            // Negate the result
+            if let Ok(solution) =  fulfill.solve(Substitution::empty()) {
+                match solution {
+                    Solution::Unique(_) => Err("refutation failed")?,
+                    Solution::Ambig(_) => return Ok(Outcome::Incomplete),
                 }
+            }
+        }
+
+        Ok(Outcome::Complete)
+    }
+
+    /// Try to fulfill all pending obligations. The returned solution will
+    /// transform `subst` substitution with the outcome of type inference by
+    /// updating the replacements it provides.
+    pub fn solve(mut self, subst: Substitution) -> Result<Solution> {
+        // We first solve all positive (to_prove) goals, since these are the
+        // only goals that can affect inference state or add new obligations.
+        let positive = self.fulfill_positive()?;
 
-                Ok(Solution::Ambig(Guidance::Unknown))
-            } else {
-                // While we failed to prove the goal, we still leared that
-                // something had to hold. Here's an example where this happens:
-                //
-                // ```rust
-                // trait Display {}
-                // trait Debug {}
-                // struct Foo<T> {}
-                // struct Bar {}
-                // struct Baz {}
-                //
-                // impl Display for Bar {}
-                // impl Display for Baz {}
-                //
-                // impl<T> Debug for Foo<T> where T: Display {}
-                // ```
-                //
-                // If we pose the goal `exists<T> { T: Debug }`, we can't say
-                // for sure what `T` must be (it could be either `Foo<Bar>` or
-                // `Foo<Baz>`, but we *can* say for sure that it must be of the
-                // form `Foo<?0>`.
-                let subst = self.infer.canonicalize(&subst);
-                Ok(Solution::Ambig(Guidance::Definite(subst.quantified)))
+        // Next we look at the goals we must refute; doing so cannot create new
+        // obligations of any form
+        let negative = self.fulfill_negative()?;
```

Comment:
Clever to stratify these like that, though probably not the most performant thing ever (i.e., we might have been able to see that we will fail to refute a "refute goal" and hence abort the process much earlier).